[FEATURE] - Integrate the output with some LLM

**Is your feature request related to a problem? Please describe.**
It would be great if we had a chance to integrate the output in case of error with some LLM such as ChatGPT or Gemini for example.

**Describe the solution you'd like**
Integrate the output with error with chatGPT or Gemini Google for example.

**Describe alternatives you've considered**
For example. Using the ChatGPT asks:

~~~
what's the error from the following SSL message: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:130)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:378)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:321)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:316)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:647)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:467)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:363)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:393)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:476)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:447)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:201)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1506)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1421)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:455)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:426)
	at io.github.gabrielpadilh4.services.SSLService.openClientSocket(SSLService.java:65)
	at io.github.gabrielpadilh4.services.SSLService.logSSLHandshake(SSLService.java:119)
	at io.github.gabrielpadilh4.commands.SSLDebugCommand.call(SSLDebugCommand.java:51)
	at io.github.gabrielpadilh4.commands.SSLDebugCommand.call(SSLDebugCommand.java:15)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at io.github.gabrielpadilh4.Main.main(Main.java:12)
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:388)
	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:271)
	at java.base/sun.security.validator.Validator.validate(Validator.java:256)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:230)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:132)
	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:631)
	... 24 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:146)
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:127)
	at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:383)
	... 29 more
~~~ 

The result is:

~~~
The error message you provided indicates a `SSLHandshakeException` with the root cause being a `SunCertPathBuilderException`. The specific error is "unable to find valid certification path to requested target." This typically occurs when the SSL/TLS client cannot establish a trusted connection to the server because it cannot validate the server's certificate.

The error is occurring during the SSL handshake process, where the client is trying to validate the server's SSL/TLS certificate using the public key infrastructure (PKI). The client is unable to build a valid certification path to the server's certificate, meaning it cannot verify the authenticity and integrity of the server's certificate.

Possible reasons for this error include:

1. **Missing or Incorrect Root Certificate:**
   - The server's SSL/TLS certificate might be signed by a certificate authority (CA) that is not recognized by the client.
   - The client may be missing the necessary root or intermediate CA certificates required to establish a chain of trust.

2. **Expired or Revoked Certificates:**
   - The server's certificate or one of the intermediate certificates in the chain may have expired.
   - The server's certificate or one of the intermediate certificates may have been revoked.

3. **Hostname Mismatch:**
   - The Common Name (CN) or Subject Alternative Name (SAN) in the server's certificate might not match the hostname to which the client is connecting.

To resolve this issue, you can consider the following steps:

- Ensure that the server's SSL/TLS certificate is valid and has not expired.
- Make sure that the certificate chain presented by the server is complete and includes all necessary intermediate certificates.
- Verify that the CA that signed the server's certificate is recognized by the client.
- Check for any hostname mismatches between the server's certificate and the hostname to which the client is connecting.

If you have control over the server, you may need to update the server's SSL/TLS certificate or ensure that the correct intermediate certificates are being presented. If you are a client connecting to a server, you may need to update your truststore with the appropriate CA certificates or resolve any hostname mismatches.
~~~

**Additional context**
It would be nice to have some parameters to allow the end user to use the LLM or don't.

--- a/README.md
+++ b/README.md
@@ -99,6 +99,50 @@ ssl-handshake-debugger -s example.com -p 443 -f output.txt client
 
 `-f output.txt`: Writes the handshake output to a file named 'output.txt'.
 
+### Using ChatGPT OpenAPI
+
+First you need to [create an API Key](https://platform.openai.com/api-keys) on OpenAPI, save it and run the following command:
+
+```{bash}
+ssl-handshake-debugger client -s self-signed.badssl.com -p 443 --openAIApiKey=YOUR_API_KEY
+```
+
+The result should be:
+
+The error message "PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target" typically occurs when the Java application is unable to establish a secure connection with a remote server due to an invalid or missing SSL certificate.
+
+To fix this issue, you can follow these steps:
+
+1. Identify the certificate authority (CA) that issued the SSL certificate for the remote server.
+2. Obtain the root and intermediate certificates from the CA.
+3. Import the root and intermediate certificates into the Java keystore.
+
+Here is a step-by-step guide:
+
+1. Identify the certificate authority (CA):
+   - Check the SSL certificate details in your web browser when accessing the remote server. Look for the "Issued by" or "Certificate Authority" information.
+   - Contact the server administrator or the CA to obtain the necessary certificates.
+
+2. Obtain the root and intermediate certificates:
+   - Visit the CA's website and search for their root and intermediate certificates.
+   - Download the certificates in PEM or DER format.
+
+3. Import the root and intermediate certificates into the Java keystore:
+   - Open a command prompt or terminal.
+   - Locate the Java installation directory and navigate to the "bin" folder.
+   - Run the following command to import the root certificate:
+     ```
+     keytool -import -alias root -keystore <path_to_java_home>/jre/lib/security/cacerts -file <path_to_root_certificate>
+     ```
+     Replace `<path_to_java_home>` with the actual path to your Java installation directory, and `<path_to_root_certificate>` with the path to the downloaded root certificate file.
+   - When prompted for the keystore password, enter the default password: "changeit".
+   - Confirm the import by typing "yes" when prompted.
+   - Repeat the above steps for each intermediate certificate, using a unique alias for each.
+
+4. Restart your Java application and try connecting to the remote server again. The error should no longer occur.
+
+Note: It's important to ensure that the root and intermediate certificates are obtained from a trusted source and are not tampered with.
+
 ## About
 
 Command line application built in Java that tests SSL/TLS handshake as client or server and prints the [javax.net.debug](https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ReadDebug.html) output.

--- a/pom.xml
+++ b/pom.xml
@@ -18,6 +18,7 @@
 		<maven-compiler-plugin-version>3.12.0</maven-compiler-plugin-version>
 		<picocli.version>4.7.5</picocli.version>
 		<appassembler.version>1.10</appassembler.version>
+		<langchain.version>0.25.0</langchain.version>
 	</properties>
 
 	<dependencies>
@@ -25,6 +26,16 @@
 			<groupId>info.picocli</groupId>
 			<artifactId>picocli</artifactId>
 			<version>${picocli.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>dev.langchain4j</groupId>
+			<artifactId>langchain4j</artifactId>
+			<version>${langchain.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>dev.langchain4j</groupId>
+			<artifactId>langchain4j-open-ai</artifactId>
+			<version>${langchain.version}</version>
 		</dependency>
 	</dependencies>
 
@@ -35,7 +46,8 @@
 				<artifactId>appassembler-maven-plugin</artifactId>
 				<version>${appassembler.version}</version>
 				<configuration>
-					<assembleDirectory>${project.build.directory}/ssl-handshake-debugger</assembleDirectory>
+					<assembleDirectory>
+						${project.build.directory}/ssl-handshake-debugger</assembleDirectory>
 					<programs>
 						<program>
 							<mainClass>io.github.gabrielpadilh4.Main</mainClass>

--- a/src/main/java/io/github/gabrielpadilh4/commands/SSLDebugCommand.java
+++ b/src/main/java/io/github/gabrielpadilh4/commands/SSLDebugCommand.java
@@ -57,6 +57,9 @@ public class SSLDebugCommand implements Callable<Integer> {
 
     @Option(names = { "--keystorePassword", "-ksp" }, description = "keystorePassword", defaultValue = "")
     private String keystorePassword;
+    
+    @Option(names = { "--openAIApiKey" }, description = "OpenAI API Key (ChatGPT)", defaultValue = "")
+    private String openAIApiKey;
 
     @Option(names = { "--all", "-a" }, description = "use javax.net.debug=all instead of javax.net.debug=ssl:handshake:verbose")
     private boolean allJavaxNetDebug;
@@ -83,6 +86,7 @@ public class SSLDebugCommand implements Callable<Integer> {
         		.keystore(keystore)
         		.keystorePassword(keystorePassword)
         		.allDebug(allJavaxNetDebug)
+        		.openAiApiKey(openAIApiKey)
         		.build();
    
         SSLService.logSSLHandshake(sslCliParams);

--- a/src/main/java/io/github/gabrielpadilh4/models/SslCliParams.java
+++ b/src/main/java/io/github/gabrielpadilh4/models/SslCliParams.java
@@ -18,6 +18,7 @@ public class SslCliParams {
     private String keystorePassword;
     private String truststore;
     private String truststorePassword;
+    private String openAiApiKey;
     private boolean allDebug;
 
     public SslCliParams() {
@@ -32,6 +33,7 @@ public class SslCliParams {
         this.keystorePassword = "";
         this.truststore = "";
         this.truststorePassword = "";
+        this.openAiApiKey = "";
         this.allDebug = false;
     }
     
@@ -48,6 +50,7 @@ public class SslCliParams {
         this.truststore = builder.truststore;
         this.truststorePassword = builder.truststorePassword;
         this.allDebug = builder.allDebug;
+        this.openAiApiKey = builder.openAiApiKey;
     }
     
     public static SslCliParamsBuilder create() {
@@ -155,4 +158,12 @@ public class SslCliParams {
     public void setAllDebug(boolean allDebug) {
         this.allDebug = allDebug;
     }
+
+	public String getOpenAiApiKey() {
+		return openAiApiKey;
+	}
+
+	public void setOpenAiApiKey(String openAiApiKey) {
+		this.openAiApiKey = openAiApiKey;
+	}
 }

--- a/src/main/java/io/github/gabrielpadilh4/models/SslCliParamsBuilder.java
+++ b/src/main/java/io/github/gabrielpadilh4/models/SslCliParamsBuilder.java
@@ -19,7 +19,13 @@ public class SslCliParamsBuilder {
 	protected String truststore;
 	protected String truststorePassword;
 	protected boolean allDebug;
+	protected String openAiApiKey;
     
+	public SslCliParamsBuilder openAiApiKey(String openAiApiKey) {
+		this.openAiApiKey = Optional.ofNullable(openAiApiKey).orElse("");
+		return this;
+	}
+	
 	public SslCliParamsBuilder mode(String mode) {
 		this.mode = Optional.ofNullable(mode).orElse("");
 		return this;

--- a/src/main/java/io/github/gabrielpadilh4/services/LLMService.java
+++ b/src/main/java/io/github/gabrielpadilh4/services/LLMService.java
@@ -1,0 +1,33 @@
+package io.github.gabrielpadilh4.services;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import dev.langchain4j.model.chat.ChatLanguageModel;
+import dev.langchain4j.model.input.Prompt;
+import dev.langchain4j.model.input.PromptTemplate;
+import dev.langchain4j.model.openai.OpenAiChatModel;
+import dev.langchain4j.model.openai.OpenAiModelName;
+
+/**
+ * @author pedro-hos@outlook.com
+ */
+public class LLMService {
+	
+	public static String askChatGPT(String error, String apiKey) {
+		PromptTemplate promptTemplate = PromptTemplate.from("How to fix {{error}}");
+		Map<String, Object> variables = new HashMap<>();
+		variables.put("error", error);
+		Prompt prompt = promptTemplate.apply(variables);
+		return withModel(apiKey).generate(prompt.text());
+	}
+	
+	private static ChatLanguageModel withModel(String apiKey) {
+		return OpenAiChatModel.builder()
+				  .apiKey(apiKey)
+				  .modelName(OpenAiModelName.GPT_3_5_TURBO)
+				  .temperature(0.3)
+				  .build();
+	}
+
+}

--- a/src/main/java/io/github/gabrielpadilh4/services/SSLService.java
+++ b/src/main/java/io/github/gabrielpadilh4/services/SSLService.java
@@ -140,7 +140,13 @@ public class SSLService {
             }
 
         } catch (Exception e) {
-            e.printStackTrace();
+        	
+        	if(!sslCliParams.getOpenAiApiKey().isBlank()) {
+        		String response = LLMService.askChatGPT(e.getMessage(), sslCliParams.getOpenAiApiKey());
+        		System.out.println(response);
+        	} else {
+        		e.printStackTrace();
+        	}
         }
     }
 }

